e2e/aws/rds_test.go

```diff
@@ -0,0 +1,224 @@
+	"github.com/gravitational/teleport/lib/srv/alpnproxy"
+	alpncommon "github.com/gravitational/teleport/lib/srv/alpnproxy/common"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func TestDatabases(t *testing.T) {
+	// get test settings
+	awsRegion := mustGetEnv(t, awsRegionEnv)
+	dbDiscoverySvcRoleARN := mustGetEnv(t, dbDiscoverySvcRoleARNEnv)
+	dbSvcRoleARN := mustGetEnv(t, dbSvcRoleARNEnv)
+	dbUser := mustGetEnv(t, dbUserEnv)
+	rdsPostgresInstanceName := mustGetEnv(t, rdsPostgresInstanceNameEnv)
+
+	cluster := createTeleportCluster(t,
+		withSingleProxyPort(t),
+		withDiscoveryService(t, "db-e2e-test", types.AWSMatcher{
+			Types:   []string{services.AWSMatcherRDS},
+			Tags:    mustGetDiscoveryMatcherLabels(t),
+			Regions: []string{awsRegion},
+			AssumeRole: &types.AssumeRole{
+				RoleARN: dbDiscoverySvcRoleARN,
+			},
+		}),
+		withDatabaseService(t, services.ResourceMatcher{
+			Labels: types.Labels{types.Wildcard: {types.Wildcard}},
+			AWS: services.ResourceMatcherAWS{
+				AssumeRoleARN: dbSvcRoleARN,
+			},
+		}),
+		withFullDatabaseAccessUserRole(t),
+	)
+
+	wantDBNames := []string{
+		rdsPostgresInstanceName,
+	}
+	// wait for the databases to be discovered
+	waitForDatabases(t, cluster.Process, wantDBNames...)
+	// wait for the database heartbeats from database service
+	waitForDatabaseServers(t, cluster.Process, wantDBNames...)
+
+	rdsPostgresInstance := tlsca.RouteToDatabase{
+		ServiceName: rdsPostgresInstanceName,
+		Protocol:    defaults.ProtocolPostgres,
+		Username:    dbUser,
+		Database:    "postgres",
+	}
+	tests := []struct {
+		name             string
+		route            tlsca.RouteToDatabase
+		testDBConnection dbConnectionTestFunc
+	}{
+		{
+			name:             "RDS postgres instance",
+			route:            rdsPostgresInstance,
+			testDBConnection: postgresConnTestFn(cluster),
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+			test.testDBConnection(t, ctx, test.route)
+		})
+	}
+}
+
+type dbConnectionTestFunc func(*testing.T, context.Context, tlsca.RouteToDatabase)
+
+// postgresConnTestFn tests connection to a postgres database.
+func postgresConnTestFn(cluster *helpers.TeleInstance) dbConnectionTestFunc {
+	return func(t *testing.T, ctx context.Context, route tlsca.RouteToDatabase) {
+		lp := startLocalALPNProxy(t, ctx, cluster, route)
+		defer lp.Close()
+
+		connString := fmt.Sprintf("postgres://%s@%v/%s",
+			route.Username, lp.GetAddr(), route.Database)
+		var pgConn *pgconn.PgConn
+		// retry for a while, the database service might need time to give
+		// itself IAM rds:connect permissions.
+		require.EventuallyWithT(t, func(c *assert.CollectT) {
+			var err error
+			pgConn, err = pgconn.Connect(ctx, connString)
+			assert.NoError(c, err)
+			assert.NotNil(c, pgConn)
+		}, time.Second*10, time.Second, "connecting to postgres")
```

Comment:
it was enough when I tried it a few times in the dev account - I tried that by deleting the inline policy that the db service auto-adds to itself and re-running.

we can extend it if we see failures, although typically the first connection will succeed since the inline policy from the first run will persist.

Reply:
> we can extend it if we see failures, although typically the first connection will succeed since the inline policy from the first run will persist.

Should we clean up the policy before each test? If we are not testing the db/cloud/iam service, might just give the `rds:connect` permissions for the specific RDS resource to the agent, instead of giving it `iam:PutRolePolicy`

Reply:
I prefer keeping things closer to real setup, even though it makes the test slows by a few seconds